nebula_ros/src/hesai/hw_monitor_wrapper.cpp

```diff
@@ -0,0 +1,399 @@
+  }
+  else
+  {
+    diagnostics.summary(diagnostic_msgs::msg::DiagnosticStatus::WARN, "No data available");
+  }
+}
+
+void HesaiHwMonitorWrapper::HesaiCheckTemperature(diagnostic_updater::DiagnosticStatusWrapper& diagnostics)
+{
+  std::scoped_lock lock(mtx_lidar_status_);
+  if (current_status_)
+  {
+    uint8_t level = diagnostic_msgs::msg::DiagnosticStatus::OK;
+    std::vector<std::string> msg;
+    for (size_t i = 0; i < std::size(current_status_->temperature); i++)
+    {
+      diagnostics.add(temperature_names_[i],
+                      GetFixedPrecisionString(current_status_->temperature[i].value() * 0.01, 3));
+    }
+    diagnostics.summary(level, boost::algorithm::join(msg, ", "));
+  }
+  else
+  {
+    diagnostics.summary(diagnostic_msgs::msg::DiagnosticStatus::WARN, "No data available");
+  }
+}
+
+void HesaiHwMonitorWrapper::HesaiCheckRpm(diagnostic_updater::DiagnosticStatusWrapper& diagnostics)
+{
+  std::scoped_lock lock(mtx_lidar_status_);
+  if (current_status_)
+  {
+    uint8_t level = diagnostic_msgs::msg::DiagnosticStatus::OK;
+    std::vector<std::string> msg;
+    diagnostics.add("motor_speed", std::to_string(current_status_->motor_speed.value()));
+
+    diagnostics.summary(level, boost::algorithm::join(msg, ", "));
+  }
+  else
+  {
+    diagnostics.summary(diagnostic_msgs::msg::DiagnosticStatus::WARN, "No data available");
+  }
+}
+
+void HesaiHwMonitorWrapper::HesaiCheckVoltageHttp(diagnostic_updater::DiagnosticStatusWrapper& diagnostics)
+{
+  std::scoped_lock lock(mtx_lidar_monitor_);
+  if (current_lidar_monitor_tree_)
+  {
+    uint8_t level = diagnostic_msgs::msg::DiagnosticStatus::OK;
+    std::vector<std::string> msg;
+    std::string key = "";
+
+    std::string mes;
+    key = "lidarInCur";
+    try
+    {
+      mes = GetPtreeValue(current_lidar_monitor_tree_.get(), "Body." + key);
+    }
+    catch (boost::bad_lexical_cast& ex)
+    {
+      level = diagnostic_msgs::msg::DiagnosticStatus::ERROR;
+      mes = MSG_ERROR + std::string(ex.what());
+    }
+    diagnostics.add(key, mes);
+    key = "lidarInVol";
+    try
+    {
+      mes = GetPtreeValue(current_lidar_monitor_tree_.get(), "Body." + key);
+    }
+    catch (boost::bad_lexical_cast& ex)
+    {
+      level = diagnostic_msgs::msg::DiagnosticStatus::ERROR;
+      mes = MSG_ERROR + std::string(ex.what());
+    }
+    diagnostics.add(key, mes);
+
+    diagnostics.summary(level, boost::algorithm::join(msg, ", "));
+  }
+  else
+  {
+    diagnostics.summary(diagnostic_msgs::msg::DiagnosticStatus::WARN, "No data available");
+  }
+}
+
+void HesaiHwMonitorWrapper::HesaiCheckVoltage(diagnostic_updater::DiagnosticStatusWrapper& diagnostics)
+{
+  std::scoped_lock lock(mtx_lidar_monitor_);
+  if (current_monitor_)
+  {
+    uint8_t level = diagnostic_msgs::msg::DiagnosticStatus::OK;
+    std::vector<std::string> msg;
+    diagnostics.add("input_voltage", GetFixedPrecisionString(current_monitor_->input_voltage.value() * 0.01, 3) + " V");
```

Comment:
Seems like these might vary a bit per sensor - should probably be parameters or macros instead of magic numbers.

Reply:
I agree. The situation is even worse with Robosense, Velodyne as their datasheets just specify a bunch of formulas with no further explanation of the constants.

As this code was taken straight from the `hw_monitor_ros_wrapper.cpp`, I would like to address this in a separate PR at a later time if that's ok.